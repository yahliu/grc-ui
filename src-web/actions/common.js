/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

import _ from 'lodash'

import GrcApolloClient from '../utils/client/apollo-client'
import { formatExpandablePolicies } from '../utils/FormatTableData'
import { buildSelfLinK } from '../utils/BuildSelfLink'
import {
  RESOURCE_RECEIVE_SUCCESS,
  REQUEST_STATUS, RESOURCE_RECEIVE_FAILURE, RESOURCE_REQUEST,
  RESOURCE_MUTATE, RESOURCE_MUTATE_SUCCESS,
  RESOURCE_MUTATE_FAILURE, AVAILABLE_FILTER_UPDATE, ACTIVE_FILTER_UPDATE,
  SECONDARY_HEADER_UPDATE, MODAL_UPDATE, POST_REQUEST, POST_RECEIVE_SUCCESS,
  POST_RECEIVE_FAILURE, PUT_REQUEST, PUT_RECEIVE_SUCCESS, PUT_RECEIVE_FAILURE,
  PATCH_REQUEST, PATCH_RECEIVE_SUCCESS, PATCH_RECEIVE_FAILURE, DEL_REQUEST,
  DEL_RECEIVE_SUCCESS, DEL_RECEIVE_FAILURE, CLEAR_REQUEST_STATUS
} from './index'

export const receiveResourceSuccess = (response, resourceType) => ({
  type: RESOURCE_RECEIVE_SUCCESS,
  status: REQUEST_STATUS.DONE,
  items: response.items,
  resourceVersion: _.get(response, 'metadata.resourceVersion'), //only supported on k8s resoruces
  resourceType
})

export const receiveResourceError = (err, resourceType) => ({
  type: RESOURCE_RECEIVE_FAILURE,
  status: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const requestResource = (resourceType) => ({
  type: RESOURCE_REQUEST,
  status: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const mutateResource = (resourceType, resourceName) => ({
  type: RESOURCE_MUTATE,
  resourceName,
  resourceType,
})

export const mutateResourceSuccess = (resourceType, resourceName) => ({
  type: RESOURCE_MUTATE_SUCCESS,
  resourceName,
  resourceType,
})

export const mutateResourceFailure = (resourceType, error) => ({
  type: RESOURCE_MUTATE_FAILURE,
  postStatus: REQUEST_STATUS.ERROR,
  err: { error },
  resourceType,
})

export const fetchResources = (resourceType, vars) => {
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.get(resourceType, vars)
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(formatExpandablePolicies(response.data.items))}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const fetchSingleResource = (resourceType, args) => {
  return (dispatch) => {
    dispatch(requestResource(resourceType))
    return GrcApolloClient.getResource(resourceType, args)
      .then(response => {
        if (response.errors) {
          return dispatch(receiveResourceError(response.errors[0], resourceType))
        }
        return dispatch(receiveResourceSuccess({items: _.cloneDeep(response.data)}, resourceType))
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const disableResource = (resourceType, namespace, name, body, resourceData, resourcePath) => (dispatch => {
  dispatch(patchResource(resourceType))
  const selfLink = buildSelfLinK(resourceData)
  return GrcApolloClient.updateResource(namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePatchError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-disable'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePatchResource(response, resourceType))
    })
})

export const enforcResource = (resourceType, namespace, name, body, resourceData, resourcePath) => (dispatch => {
  dispatch(patchResource(resourceType))
  const selfLink = buildSelfLinK(resourceData)
  return GrcApolloClient.updateResource(namespace, name, body, selfLink, resourcePath)
    .then(response => {
      if (response.errors) {
        return dispatch(receivePatchError(response.errors[0], resourceType))
      } else {
        dispatch(updateModal({open: false, type: 'resource-enforce'}))
      }
      dispatch(fetchResources(resourceType))
      return dispatch(receivePatchResource(response, resourceType))
    })
})

export const removeResource = (resourceType, resourceData) => async dispatch => {
  dispatch(delResource(resourceType))
  const selfLink = buildSelfLinK(resourceData)
  try {
    const response = await GrcApolloClient.remove(resourceData, selfLink)
    if (response.errors) {
      return dispatch(receiveDelError(response.errors, resourceType))
    } else {
      dispatch(receiveDelResource(response, resourceType, resourceData))
    }
    dispatch(fetchResources(resourceType))
  } catch (err) {
    return dispatch(receiveDelError(err, resourceType))
  }
}

export const updateActiveFilters = (activeFilters) => ({
  type: ACTIVE_FILTER_UPDATE,
  activeFilters,
})

export const updateAvailableFilters = (availableFilters) => ({
  type: AVAILABLE_FILTER_UPDATE,
  availableFilters
})

export const updateSecondaryHeader = (title, tabs, breadcrumbItems, links, description, information) => ({
  type: SECONDARY_HEADER_UPDATE,
  title,
  tabs,
  breadcrumbItems,
  links,
  description,
  information
})

export const updateModal = (data) => ({
  type: MODAL_UPDATE,
  data
})

export const postResource = (resourceType) => ({ // TODO: Consider renaming
  type: POST_REQUEST,
  postStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePostResource = (item, resourceType) => ({
  type: POST_RECEIVE_SUCCESS,
  postStatus: REQUEST_STATUS.DONE,
  resourceType: item.kind || resourceType,
  item
})

export const receivePostError = (err, resourceType) => ({
  type: POST_RECEIVE_FAILURE,
  postStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const putResource = (resourceType) => ({ // TODO: Consider renaming
  type: PUT_REQUEST,
  putStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePutResource = (item, resourceType) => {
  return ({
    type: PUT_RECEIVE_SUCCESS,
    putStatus: REQUEST_STATUS.DONE,
    resourceType: item.kind || resourceType,
    item
  })
}

export const receivePutError = (err, resourceType) => ({
  type: PUT_RECEIVE_FAILURE,
  putStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const patchResource = (resourceType) => ({ // TODO: Consider renaming
  type: PATCH_REQUEST,
  patchStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receivePatchResource = (item, resourceType) => {
  return ({
    type: PATCH_RECEIVE_SUCCESS,
    patchStatus: REQUEST_STATUS.DONE,
    resourceType: item.kind || resourceType,
    item
  })
}

export const receivePatchError = (err, resourceType) => ({
  type: PATCH_RECEIVE_FAILURE,
  patchStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const delResource = (resourceType) => ({ // TODO: Consider renaming
  type: DEL_REQUEST,
  delStatus: REQUEST_STATUS.IN_PROGRESS,
  resourceType
})

export const receiveDelResource = (item, resourceType, resource) => ({
  type: DEL_RECEIVE_SUCCESS,
  delStatus: REQUEST_STATUS.DONE,
  resourceType: item.kind || resourceType,
  item,
  resource
})

export const receiveDelError = (err, resourceType) => ({
  type: DEL_RECEIVE_FAILURE,
  delStatus: REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const clearRequestStatus = (resourceType) => ({
  type: CLEAR_REQUEST_STATUS,
  resourceType: resourceType
})

export const createResources = (resourceType, resourceJson) => {
  return (dispatch) => {
    dispatch(mutateResource(resourceType))
    return GrcApolloClient.createResources(resourceJson)
      .then(result => {
        if (result.data.createResources.errors && result.data.createResources.errors.length > 0){
          dispatch(mutateResourceFailure(resourceType, result.data.createResources.errors[0]))
        } else {
          dispatch(mutateResourceSuccess(resourceType))
        }
        return result
      })
  }
}

export const createAndUpdateResources = (resourceTypes, createList, updateList) => {
  return (dispatch) => {
    resourceTypes.forEach((resourceType) => {
      dispatch(mutateResource(resourceType))
    })
    return GrcApolloClient.createAndUpdateResources(createList, updateList)
      .then(result => {
        const errors = {
          Policy: {
            resourceType: 'Policy',
            error: '',
          },
          PlacementRule: {
            resourceType: 'PlacementRule',
            error: '',
          },
          PlacementBinding: {
            resourceType: 'PlacementBinding',
            error: '',
          }
        }
        if (_.get(result, 'data.createAndUpdateResources.create.errors') &&
          _.get(result, 'data.createAndUpdateResources.create.errors.length') > 0){
          _.get(result, 'data.createAndUpdateResources.create.errors', []).forEach((error) => {
            errors[error.kind].error = error.message
          })
        }
        if (_.get(result, 'data.createAndUpdateResources.update.errors') &&
          _.get(result, 'data.createAndUpdateResources.update.errors.length') > 0){
          _.get(result, 'data.createAndUpdateResources.update.errors', []).forEach((error) => {
            errors[error.kind].error = error.message
          })
        }
        let errored = false
        Object.keys(errors).forEach((key) => {
          const resp = errors[key]
          if (resp.error !== '') {
            errored = true
            dispatch(mutateResourceFailure(resp.resourceType, { message: resp.error }))
          }
        })
        if (!errored) {
          resourceTypes.forEach((resourceType) => {
            dispatch(mutateResourceSuccess(resourceType))
          })
        }
        return result
      })
  }
}

export const createResource = (resourceType, variables) => {
  return (dispatch) => {
    dispatch(postResource(resourceType))
    return GrcApolloClient.createResource(resourceType, variables)
      .then(response => {
        if (response.errors) {
          return dispatch(receivePostError(response.errors[0], resourceType))
        }

        return dispatch(receivePostResource(_.cloneDeep(response.data.setHelmRepo), resourceType))
      })
      .catch(err => dispatch(receivePostError(err, resourceType)))
  }
}

export const copyAnsibleSecret = (name, namespace, targetNamespace) => {
  return (dispatch) => {
    return GrcApolloClient.copyAnsibleSecret(name, namespace, targetNamespace)
      .then(response => {
        return response
      })
      .catch(err => dispatch(receivePostError(err)))
  }
}

export const getPolicyAutomation = (namespace) => {
  return (dispatch) => {
    return GrcApolloClient.getPolicyAutomation(namespace)
      .then(response => {
        return response
      })
      .catch(err => dispatch(receivePostError(err)))
  }
}

export const modifyPolicyAutomation = (poliyAutomationJSON, action) => {
  return (dispatch) => {
    return GrcApolloClient.modifyPolicyAutomation(poliyAutomationJSON, action)
      .then(response => {
        return response
      })
      .catch(err => dispatch(receivePostError(err)))
  }
}
