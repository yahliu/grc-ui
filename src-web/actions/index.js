/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

/*
 * action types
 *
 * action name: <NOUN>_<VERB>
 */

export const DEL_RECEIVE_FAILURE = 'DEL_RECEIVE_FAILURE'
export const DEL_RECEIVE_SUCCESS = 'DEL_RECEIVE_SUCCESS'
export const DEL_REQUEST = 'DEL_REQUEST'
export const CLEAR_REQUEST_STATUS = 'CLEAR_REQUEST_STATUS'
export const MODAL_UPDATE = 'MODAL_UPDATE'
export const NAV_MODIFY = 'NAV_MODIFY'
export const NAV_RECEIVE_SUCCESS = 'NAV_RECEIVE_SUCCESS'
export const POST_RECEIVE_FAILURE = 'POST_RECEIVE_FAILURE'
export const POST_RECEIVE_SUCCESS = 'POST_RECEIVE_SUCCESS'
export const POST_REQUEST = 'POST_REQUEST'
export const PUT_RECEIVE_FAILURE = 'PUT_RECEIVE_FAILURE'
export const PUT_RECEIVE_SUCCESS = 'PUT_RECEIVE_SUCCESS'
export const PUT_REQUEST = 'PUT_REQUEST'
export const PATCH_RECEIVE_FAILURE = 'PATCH_RECEIVE_FAILURE'
export const PATCH_RECEIVE_SUCCESS = 'PATCH_RECEIVE_SUCCESS'
export const PATCH_REQUEST = 'PATCH_REQUEST'
export const RESOURCES_FETCH_REQUEST_LOADING = 'RESOURCE_FETCH_REQUEST_LOADING'
export const RESOURCES_FETCH_REQUEST_SUCCESS = 'RESOURCES_FETCH_REQUEST_SUCCESS'
export const RESOURCE_MUTATE = 'RESOURCE_MUTATE'
export const RESOURCE_MUTATE_FAILURE = 'RESOURCE_MUTATE_FAILURE'
export const RESOURCE_MUTATE_SUCCESS = 'RESOURCE_MUTATE_SUCCESS'
export const RESOURCE_RECEIVE_FAILURE = 'RESOURCE_RECEIVE_FAILURE'
export const RESOURCE_RECEIVE_SUCCESS = 'RESOURCE_RECEIVE_SUCCESS'
export const RESOURCE_FILTERS_RECEIVE_SUCCESS = 'RESOURCE_FILTERS_RECEIVE_SUCCESS'
export const RESOURCE_FILTERS_UPDATE = 'RESOURCE_FILTERS_UPDATE'
export const RESOURCE_REQUEST = 'RESOURCE_REQUEST'
export const USER_ACCESS_FAILURE = 'USER_ACCESS_FAILURE'
export const USER_ACCESS_SUCCESS = 'USER_ACCESS_SUCCESS'
export const SECONDARY_HEADER_UPDATE = 'SECONDARY_HEADER_UPDATE'
export const RESOURCE_TOOLBAR_UPDATE = 'RESOURCE_TOOLBAR_UPDATE'
export const ACTIVE_FILTER_UPDATE = 'ACTIVE_FILTER_UPDATE'
export const AVAILABLE_FILTER_UPDATE = 'AVAILABLE_FILTER_UPDATE'

/*
 * other constants specific to particular actions
 */

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE'
}

export const USER_LOGIN_STATUS = {
  LOGGED_IN: 'LOGGED_IN',
  LOGGED_OUT: 'LOGGED_OUT'
}

export const REQUEST_STATUS = {
  INCEPTION: 'INCEPTION',
  DONE: 'DONE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR'
}

export const PAGE_SIZES = {
  DEFAULT: 10,
  VALUES: [5, 10, 20, 50, 75, 100]
}

export const SORT_DIRECTION_ASCENDING = 'asc'
export const SORT_DIRECTION_DESCENDING = 'desc'
