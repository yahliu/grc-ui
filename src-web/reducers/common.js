/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

/*
Since these selectors are common (i.e used across different components and different parts of the store)
we have to create unique selectors each invocation and the selectors need additional metata data sent
though props that indicate which part of the store it should select from.

//See https://github.com/reactjs/reselect#sharing-selectors-with-props-across-multiple-components
//See https://github.com/reactjs/reselect#q-can-i-share-a-selector-across-multiple-components

The selector pattern is an abstraction that standardizes an applicationâ€™s store querying logic.
It is simple: for any part of the store that an application needs access to, define a function that
when given the full store, returns the desired part (or derivation) of the store.
*/

import _ from 'lodash'
import { RESOURCE_TYPES } from '../utils/constants'
import {
  PAGE_SIZES, CLEAR_REQUEST_STATUS, POST_REQUEST, PUT_REQUEST,
  PATCH_REQUEST, ACTIVE_FILTER_UPDATE, AVAILABLE_FILTER_UPDATE, REQUEST_STATUS, POST_RECEIVE_SUCCESS,
  PUT_RECEIVE_SUCCESS, PATCH_RECEIVE_SUCCESS, DEL_RECEIVE_SUCCESS,
  POST_RECEIVE_FAILURE, PUT_RECEIVE_FAILURE, PATCH_RECEIVE_FAILURE,
  SECONDARY_HEADER_UPDATE, RESOURCE_REQUEST, RESOURCE_RECEIVE_SUCCESS,
  RESOURCE_RECEIVE_FAILURE, RESOURCE_MUTATE, RESOURCE_MUTATE_FAILURE,
  RESOURCE_MUTATE_SUCCESS
} from '../actions'

export const INITIAL_STATE = {
  items: [],
  itemsPerPage: PAGE_SIZES.DEFAULT,
  page: 1,
  search: '',
  sortColumn: undefined,
  status: REQUEST_STATUS.INCEPTION,
  patchStatus: undefined,
  patchErrorMsg: '',
  putStatus: undefined,
  putErrorMsg: '',
  postStatus: undefined,
  postStatusCode: undefined,
  postErrorMsg: '',
  pendingActions: [],
}

export const resourceToolbar = (state = {}, action) => {
  switch (action.type) {
  case ACTIVE_FILTER_UPDATE:
    return Object.assign({}, state, {
      activeFilters: action.activeFilters,
    })
  case AVAILABLE_FILTER_UPDATE:
    return Object.assign({}, state, {
      availableFilters: action.availableFilters,
    })
  default:
    return state
  }
}

export const secondaryHeader = (state = {title: '', tabs: [], breadcrumbItems: [], links: [], description: {}, information:{}}, action) => {
  switch (action.type) {
  case SECONDARY_HEADER_UPDATE:
    return Object.assign({}, state, {
      title: action.title,
      tabs: action.tabs,
      breadcrumbItems: action.breadcrumbItems,
      links: action.links,
      description: action.description,
      information: action.information
    })
  default:
    return state
  }
}

export const resourceReducerFunction = (state = INITIAL_STATE, action) => {
  let items,index
  switch (action.type) {
    case RESOURCE_REQUEST:
      return Object.assign({}, state, {
        status: REQUEST_STATUS.IN_PROGRESS
      })
    case RESOURCE_RECEIVE_SUCCESS: {
     let pageNumber = 1
      if (action.items.length !== 0 && state.page && state.itemsPerPage) {
        if (action.items.length > state.itemsPerPage * (state.page - 1)) {
          pageNumber = state.page
        } else {
          pageNumber = state.page - 1
        }
      }
      return Object.assign({}, state, {
        status: REQUEST_STATUS.DONE,
        items: action.items,
        page: pageNumber,
        resourceVersion: action.resourceVersion
      })
    }
    case RESOURCE_RECEIVE_FAILURE:
      return Object.assign({}, state, {
        status: REQUEST_STATUS.ERROR,
        err: action.err
      })
    case POST_REQUEST:
      return Object.assign({}, state, {
        postStatus: REQUEST_STATUS.IN_PROGRESS
      })
    case POST_RECEIVE_SUCCESS:
      items = state.items.slice(0)
      if (action.item.length > 0) { // if returned as an array due to making async calls, push action.item elements to the items array
        action.item.forEach(el => {
          items.push(el)
        })
      } else {
        items.push(action.item)
      }
      return Object.assign({}, state, {
        items: items,
        postStatus: REQUEST_STATUS.DONE
      })
    case POST_RECEIVE_FAILURE:
      return Object.assign({}, state, {
        postStatus: REQUEST_STATUS.ERROR,
        postStatusCode: action.err.error && action.err.error.response && action.err.error.response.status,
        postErrorMsg: action.err.error && action.err.error.message
      })
    case PUT_REQUEST:
      return Object.assign({}, state, {
        putStatus: REQUEST_STATUS.IN_PROGRESS
      })
    case PUT_RECEIVE_SUCCESS:
      return Object.assign({}, state, {
        putStatus: REQUEST_STATUS.DONE,
      })
    case PUT_RECEIVE_FAILURE:
      return Object.assign({}, state, {
        putStatus: REQUEST_STATUS.ERROR,
        putErrorMsg: action.err.error ? action.err.error.message : action.err.message
      })
    case PATCH_REQUEST:
      return Object.assign({}, state, {
        patchStatus: REQUEST_STATUS.IN_PROGRESS
      })
    case PATCH_RECEIVE_SUCCESS:
      return Object.assign({}, state, {
        patchStatus: REQUEST_STATUS.DONE,
      })
    case PATCH_RECEIVE_FAILURE:
      return Object.assign({}, state, {
        patchStatus: REQUEST_STATUS.ERROR,
        patchErrorMsg: action.err.error ? action.err.error.message : action.err.message
      })
    case CLEAR_REQUEST_STATUS:
      return Object.assign({}, state, {
        mutateStatus: undefined,
        mutateErrorMsg: undefined,
        postStatus: undefined,
        postStatusCode: undefined,
        postErrorMsg: undefined,
        putStatus: undefined,
        putErrorMsg: undefined,
        patchStatus: undefined,
        patchErrorMsg: undefined
      })
    case RESOURCE_MUTATE:
      return Object.assign({}, state, {
        mutateStatus: REQUEST_STATUS.IN_PROGRESS,
        mutateErrorMsg: null,
        pendingActions: [...state.pendingActions, { name: action.resourceName, action: RESOURCE_MUTATE }]
      })
    case RESOURCE_MUTATE_FAILURE:
      return Object.assign({}, state, {
        mutateStatus: REQUEST_STATUS.ERROR,
        mutateErrorMsg: action.err.message || action.err.error && (action.err.error.message ||
          (action.err.error.data && action.err.error.data.Message)),
        pendingActions: state.pendingActions.filter(r => r && r.name !== action.resourceName),
      })
    case RESOURCE_MUTATE_SUCCESS:
      return Object.assign({}, state, {
        mutateStatus: REQUEST_STATUS.DONE,
        pendingActions: state.pendingActions.filter(r => r && r.name !== action.resourceName),
      })
    case DEL_RECEIVE_SUCCESS:
      {
        items = [...state.items]
        switch (action.resourceType) {
        case RESOURCE_TYPES.POLICIES_BY_POLICY:
        case RESOURCE_TYPES.POLICY: {
          const policy = _.get(action, 'resource')
          index = _.findIndex(items, { 'name':policy.name, 'namespace':policy.namespace })
          break
        }
        default:
          index = _.findIndex(items, o => _.get(o, 'Name') === _.get(action, 'resourceName'))
          break
        }
        if(index > -1) {
          items.splice(index, 1)
          return Object.assign({}, state, {
            items: items
          })
        }
        return state
      }
    default:
      return state
  }
}

/**
 * A common higher order reducer for resources
 * Follows the Redux pattern described here: http://redux.js.org/docs/recipes/reducers/ReusingReducerLogic.html#customizing-behavior-with-higher-order-reducers
 * **/
export const createResourceReducer = (reducerFunction, reducerPredicate) => {
  return (state, action) => {
    const isInitializationCall = state === undefined
    const shouldRunWrappedReducer = reducerPredicate(action) || isInitializationCall
    return shouldRunWrappedReducer ? reducerFunction(state, action) : state
  }
}
