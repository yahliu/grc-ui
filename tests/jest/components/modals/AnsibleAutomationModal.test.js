/* Copyright (c) 2021 Red Hat, Inc. */
/* Copyright Contributors to the Open Cluster Management project */

'use strict'

import React from 'react'
import { AnsibleAutomationModal } from '../../../../src-web/components/modals/AnsibleAutomationModal'
import * as reducers from '../../../../src-web/reducers'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import thunkMiddleware from 'redux-thunk'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { AcmAlert } from '@open-cluster-management/ui-components'
import { Spinner } from '@patternfly/react-core'

// import {
//   disableResourceModalLabels
// } from './ModalsTestingData'
import { REQUEST_STATUS } from '../../../../src-web/actions/index'

describe('AnsibleAutomationModal component', () => {
  it('Disable modal renders as expected', () => {
    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const component = shallow(
      <AnsibleAutomationModal
        type='resource-disable'
        handleClose={jest.fn()}
        handleSubmit={jest.fn()}
        locale={'en'}
        open={true}
        store={store}
      />
    )
    expect(toJson(component)).toMatchSnapshot()
  })

  it('Modal renders without data', () => {

    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const component = shallow(
      <AnsibleAutomationModal
        label={{}}
        locale={'en'}
        open={true}
        store={store}
      />
    )
    expect(toJson(component)).toMatchSnapshot()
  })

  it('Modal returns spinner when loading', () => {

    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const component = shallow(
      <AnsibleAutomationModal
        label={{}}
        locale={'en'}
        open={true}
        reqStatus={REQUEST_STATUS.IN_PROGRESS}
        store={store}
      />
    )
    expect(component.find(Spinner)).toHaveLength(1)
    expect(toJson(component)).toMatchSnapshot()
  })

  it('Modal returns notification on error', () => {

    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
    const middleware = [thunkMiddleware]
    const store = createStore(combineReducers(reducers), composeEnhancers(
      applyMiddleware(...middleware)
    ))
    const component = shallow(
      <AnsibleAutomationModal
        label={{}}
        locale={'en'}
        open={true}
        reqStatus={REQUEST_STATUS.ERROR}
        reqErrorMsg='There was an error.'
        store={store}
      />
    )
    expect(component.find(AcmAlert)).toHaveLength(1)
    expect(toJson(component)).toMatchSnapshot()
  })
})
