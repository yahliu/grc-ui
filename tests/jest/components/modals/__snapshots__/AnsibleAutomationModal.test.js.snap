// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnsibleAutomationModal component Disable modal renders as expected 1`] = `
<Query
  pollInterval={0}
  query={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "getAnsibleCredentials",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "ansibleCredentials",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "namespace",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "host",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "token",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 118,
        "start": 0,
      },
    }
  }
  variables={Object {}}
>
  <Component />
</Query>
`;

exports[`AnsibleAutomationModal component Modal renders without data 1`] = `
<Query
  pollInterval={0}
  query={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "getAnsibleCredentials",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "ansibleCredentials",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "namespace",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "host",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "token",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 118,
        "start": 0,
      },
    }
  }
  variables={Object {}}
>
  <Component />
</Query>
`;
